### PR TITLE
Deprecate kube-lego and document cert-manager usage

### LINKED ISSUE #475: Move from kube-lego to cert-manager
From @willingc in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/437#issuecomment-361342249:

> We should begin looking at cert-manager as kube-lego is deprecated and will not support kubernetes > 1.8. We're currently using Kubernetes 1.8.6.

> Here's a guide for migrating to cert-manager:
> https://github.com/jetstack/cert-manager/blob/master/docs/user-guides/migrating-from-kube-lego.md

> Possibly related DNS issue jetstack/kube-lego#297

We should switch to cert-manager for 0.7, and hopefully do so in a way that doesn't interrupt user service!

### PR DESCRIPTION
**Pull request by clkao**
Note that gce ingress is currently having a regression preventing first-time cert-manager request fail (when the tls secret is unset): https://github.com/kubernetes/ingress-gce/pull/388, we might want to hold off until gce is fixed to avoid confusion.

Closes #475

### CODE REVIEW

**Comment by consideRatio** [comment]
Woooooo!! I'm enthusiastic about this :D

File: doc/source/advanced.md [hunk 1]
```diff
@@ -41,25 +41,46 @@ Note that Nginx maintains two additional ingress controllers.
...
    `<hostname>` from a browser should route traffic to the hub.
-2. Install & configure kube-lego using the
-   [kube-lego helm-chart](https://github.com/kubernetes/charts/tree/master/stable/kube-lego).
-   Remember to change `config.LEGO_EMAIL` and `config.LEGO_URL` at the least.
-3. Add an annotation + TLS config to the ingress so kube-lego knows to get certificates for
+
+2. Install & configure cert-manager using the
+   [cert-manager helm-chart](https://github.com/kubernetes/charts/tree/master/stable/cert-manager) with ingressShim enabled: `--set ingressShim.defaultIssuerName=letsencrypt-prod --set ingressShim.defaultIssuerKind=ClusterIssuer`.
+
+3. Create the default ClusterIssuer:
```
[the remark this answers is not in the document]
Comment (consideRatio) [reply doc/source/advanced.md]: We can I think - we can declare it as a dependency.

I've done that for the z2jh-chart itself and built on top of it.

For reference: https://github.com/consideRatio/z2jh-extended/blob/master/z2jh-extended/requirements.yaml

[the remark this answers is not in the document]
Comment (clkao) [reply doc/source/advanced.md]: It is very likely the users already run or will run cert-manager cluster-wide, if using it. having it as a hard dependency and the issuer managed by us seems to complicate things.

[the remark this answers is not in the document]
Comment (clkao) [reply doc/source/advanced.md]: @betatim just added the full command

File: doc/source/security.md [hunk 2]
```diff
@@ -31,25 +31,7 @@ on HTTPS security see the certificates section of [this blog post](https://blog.
...
-     https:
-       hosts:
-         - <your-domain-name>
-       letsencrypt:
-         contactEmail: <your-email-address>
-   ```
-
-2. Apply the config changes by running `helm upgrade ...`
-3. Wait for about a minute, now your hub should be HTTPS enabled!
+Refer to the Advanced Topics section for Automatic HTTPS with cert-manager and [Let's Encrypt](https://letsencrypt.org/).
```
Comment (minrk) [inline doc/source/security.md]: I want to make sure that enabling letsencrypt remains the simplest and first recommended way to enable HTTPS. It shouldn't be moved to advanced topics. This https-enabling config that faces the users should ideally stay the same and the kube-lego->certmanager transition should be hidden from the user as much as possible.

What's the impediment to doing this now?

**Comment by jhamman** [comment]
Is there a specific blocker to moving this PR forward? We're running into some related challenges that we'd like to use cert-manager for.

**Comment by consideRatio** [comment]
[kubucations's videos about cert-manager + lets encrypt](https://www.youtube.com/watch?v=etC5d0vpLZE&t=0s&index=5&list=PLShDm2AZYnK3cWZpOjV7nOpL7plH2Ztz0) was great for me to watch, that made me grasp a lot of things about cert-manager.

**I have now tried cert-manager, and will summarize what I've picked up:**
1. Have a cert-manager in your cluster, some namespace to look for issuer/certificate resources all over.
2. Have a issuer and certificate resource in some namespace
3. cert-manager creates a pod, updates an existing service to redirect traffic from a certain port to that pod in order to meet a "i can confirm i own this domain challenge" by lets encrypt.
4. When that is done, everything is returned to normal, and cert-manager will do it again when renewal is needed.

**Conclusion:**
1. our chart can provide an option to install the needed issuer / certificate with a configurable switch, allowing cert-manager to handle the TLS setup if cert-manager is running in the cluster.
2. i've learned that one can also [optionally install chart dependencies](https://github.com/helm/helm/blob/master/docs/charts.md#tags-and-condition-fields-in-requirementsyaml)

**Suggestions:**
1. We let the user declare if they would want a certificate/issuer created to be utilized by some cert-manager
2. We let the user declare if they would want a cert-manager chart installed as a dependency

---

@betatim @minrk @yuvipanda, what do you think? Especially curious regarding the optional chart dependency solution.

File: doc/source/security.md (hunk 2 above)
Comment (betatim) [reply doc/source/security.md]: Addressing this is mandatory and in my eyes what needs to be done before we can merge this.

**Comment by clkao** [comment]
An optional dependency makes it easier for people to get started, by just editing a single values file, but that does mean we have to support it and follow upstream. IMO this is more worthwhile when the dependencies are very coupled, such as a database that you want to share secrets with.

For cert manager i think asking users to run an additional `helm install` and configure the issuer should be fine.

Another scenario: if we make this an optional dependency and do not support all options, and the users wish to migrate to a standalone cert-manager (for example if users decide to switch from http01 to dns01 domain validation), we'll need to figure out how to disconnect the helm dependency and preventing the old tls secret being deleted in the process.

File: doc/source/security.md (hunk 2 above)
Comment (clkao) [reply doc/source/security.md]: Happy to add the simple command back to the security section.

**Comment by consideRatio** [comment]
@clkao it will be fine to loose the tls secret though right? It will just force a renewal ahead of time i think.

Adding cert-manager as an optional dependency, makes the users able to control it directly through provided `config.yaml`. So we would support all options, I think.

Example of `config.yaml` below:

```yaml
proxy:
  secretToken: asdfasdfasdf

cert-manager:
  someCertManagerKey: andValue
```

**Comment by minrk** [comment]
I think an additional `helm install` is fine, but right now, it requires not just a helm install, but an additional raw `kubectl create` command for the issuer (which I don't see in the docs), with coordination of values that must occur in the cert-manager install, the issuer object, and the hub ingress (maybe?). I don't think that's clear enough yet for z2jh. The automatic https deployment should be as easy as possible, and not require referring to any advanced instructions. Users should be able to deploy a cluster with https, with only:

- specify hosts, lestencrypt email in config.yaml
- possible additional verbatim copy/paste, such as `helm install stable/cert-manager`

I'd really like to avoid any `kubectl create` calls, if we can, since that doesn't get cleaned up by things like `helm del`.

What about:

1. create the default issuer in our chart by default `if https.hosts` (harmless if unused, right?)
2. include the appropriate `helm install` command to deploy cert-manager to work with our issuer

Alternately, and I don't know what the up/downsides are, ship a cert-manager in our chart. What are the downsides, for instance, of having multiple cert-manager instances in a cluster? Is it a problem? If not, having the simples autohttps config deploy cert-manager, but make sure the chart works with a central cert-manager might be ideal.

**Comment by consideRatio** [comment]
@minrk to make cert-manager get a secret with TLS certificates etc in it for a certain domain, you need both an Issuer (or ClusterIssuer) and a Certificate resource

You can automate the creation of a Certificate resource if you have an Ingress resource with a tls field in its spec though, because cert-manager will find it and use it to create the Certificate.

To conclude so far, we need an Issuer of assume the usage of a predefined ClusterIssuer, and we need a Certificate or a Ingress resource with tls fields in its spec. Then we get ourselves a TLS-secret automatically created for us to consume.

If we declare an ingress object though, that specific cloud provider needs to create an associated LoadBalancer and be able to terminate TLS etc. These things are currently in flux, and many different cloud providers have implemented this to varying degree.

The GitLab helm chart deploys their own ingress controller (pods are spawned and running) in the namespace along with their deployment, that makes them work better on all kinds of cloud provider.

---

My ideas right now.
- Not happy about deploying the kubernetes official ingress controller called `ingress-nginx` if we need to have the proxy pod running anyhow.
- So then we should only need a cert-manager installed.
- Then our chart should support:
  1. Using a predefined ClusterIssuer
  2. Dynamic optional creation of an namespaced Issuer (think of these as representing an letsencrypt account + letsencrypt server)
  3. Dynamic optional creation of a namespaced Certificate
  4. Proxy to consume a secret of a given name (it already does!)
  5. The Ingress object to support custom annotation (it already does!)

Hmmm... Something like this.

**Comment by yuvipanda** [comment]
My general hope is to keep stringing kube-lego around until we get the traefik based proxy going - although I'll defer to others wether this is tenable. That should simplify everything here. In the meantime, additional docs on using cert-manager instead of our automatic issuer seems like a good thing to do?

**Comment by yuvipanda** [comment]
https://github.com/jupyterhub/outreachy/blob/master/ideas/traefik-jupyterhub-proxy.rst has info on the traefik proxy.

**Comment by stevegore** [comment]
I'm currently having issues with `kube-lego` with very straight-forward implementation of the chart. Any chance this PR could be merged soon? @clkao @consideRatio

**Comment by betatim** [comment]
The blocker is mentioned in this comment https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/755#discussion_r202441645. The user facing config should remain as simple as it is today and be unchanged for the "just use letsencrypt" use case.

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/server-becomes-stale-in-a-matter-of-days-even-hours/1920/7

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/how-to-get-csr-for-ssl-cert/2505/2

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/how-to-get-csr-for-ssl-cert/2505/2

**Comment by supern8ent** [comment]
Just want to point out that when you run `helm install --name my-release stable/cert-manager` per the version of [cert-manager](https://github.com/helm/charts/tree/master/stable/cert-manager) linked in the instructions, it deploys but also warns,

```
**This Helm chart is deprecated**.
All future changes to the cert-manager Helm chart should be made in the
official repository: https://github.com/jetstack/cert-manager/tree/master/deploy.
The latest version of the chart can be found on the Helm Hub: https://hub.helm.sh/charts/jetstack/cert-manager.
```

**Comment by supern8ent** [comment]
I just wanted to point out that running  
```
helm install stable/cert-manager \
  --name cert-manager \
  --namespace cert-manager \
  --set ingressShim.defaultIssuerName=letsencrypt-prod \
  --set ingressShim.defaultIssuerKind=ClusterIssuer
```  
deploys but also warns,  
```
**This Helm chart is deprecated**.
All future changes to the cert-manager Helm chart should be made in the
official repository: https://github.com/jetstack/cert-manager/tree/master/deploy.
The latest version of the chart can be found on the Helm Hub: https://hub.helm.sh/charts/jetstack/cert-manager.
```

**Comment by supern8ent** [comment]
I tried the instructions in `Advanced.md` and it doesn't seem to be working. The cert-manager logs suggest that that part is working, but the system continues to serve on http, not https. 

As I read the instructions, all that is needed in config.yaml is 

```
proxy:
  secretToken: <token>
ingress:
  annotations:
    kubernetes.io/tls-acme: "true"
  tls:
   - hosts:
      - <hostname>
     secretName: jupyterhub-tls
```

I tried that. I also tried adding 

```
https:
  enabled: true
  type: manual
```

And I tried adding 
```
  enabled: true
  hosts:
    - <mydomain.com>
```
in the ingress section.

**Comment by abal09** [comment]
> I tried the instructions in `Advanced.md` and it doesn't seem to be working. The cert-manager logs suggest that that part is working, but the system continues to serve on http, not https.
> 
> As I read the instructions, all that is needed in config.yaml is
> 
> ```
> proxy:
>   secretToken: <token>
> ingress:
>   annotations:
>     kubernetes.io/tls-acme: "true"
>   tls:
>    - hosts:
>       - <hostname>
>      secretName: jupyterhub-tls
> ```
> 
> I tried that. I also tried adding
> 
> ```
> https:
>   enabled: true
>   type: manual
> ```
> 
> And I tried adding
> 
> ```
>   enabled: true
>   hosts:
>     - <mydomain.com>
> ```
> 
> in the ingress section.

```
proxy:
  secretToken: <token>
ingress:
  enabled:true
  annotations:
    kubernetes.io/tls-acme: "true"
  tls:
   - hosts:
      - <hostname>
     secretName: jupyterhub-tls
```

Add `enabled=true` underneath the ingress

**Comment by akaszynski** [comment]
It should be noted that the ``--name`` flag is now mandatory for helm3

**Comment by yuvipanda** [comment]
We no longer use kube-lego, with #1539. I've personally never been able to get cert-manager to work despite trying several times - it feels very heavy for our common use case. How do people think we should proceed with this PR now?

**Comment by stevegore** [comment]
I use cert-manager in production across a number of sites and it's been great, especially in the last few releases.

**Comment by consideRatio** [comment]
While I think cert-manager is a great utility, I'm less happy about embedding a dependency on it in this helm chart, this is an evolved opinion. I have learned from experience that it is hard to install/upgrade CRDs and we would run into that using cert-manager as a dependency.

@yuvipanda have implemented a great replacent of kube-lego that is lightweight and suitable for anyone wanting to enable HTTPs quickly, that also doesn't exclude use of cert-manager if one have that utility in the cluster.

@clkao do you agree to close this PR? I'm very thankful for your work on it, and it made us have relevant discussion and helped me learn a lot more about this topic.

**Comment by consideRatio** [comment]
Closing as 0.9.0-beta.3 has the new autohttps with a traefik pod that acquires a cert and syncs it to a k8s secret.

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/jupyterhub-upgrade-steps-for-https-acmev2/4029/1

**Pull request closed without merging**